site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,576 @@
+
+![cow decorated](/blog/2020-06-10-opentracing/cowdecorated.png) 
+
+We wrapped each method of `Cow` in a chain of additional functions `f(x) = y` became `f(x) = a(b(y))`, each function having it's own responsibility.
+
+If we apply the same pattern to our problem, we can decorate all of `mattermost-server` API calls with OpenTracing, without actually modifying the functions themselves!
+
+Implementing such functionality in other, dynamic languages is rather trivial. For example, here's how JavaScript handles it. Given a simple `Cow` object:
+
+```javascript
+const cow = {
+  feed: function(x) {
+    return `Ate for ${x} seconds!`
+  },
+  speak: function(x) {
+    return `${"Moo ".repeat(x)}!`
+  }
+}
+
+console.log(cow.feed(20))
+console.log(cow.speak(3))
+```
+
+We can wrap it in a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy):
+
+```javascript
+const tracerHandler = {
+  get: function(target, prop, receiver) {
+    if (typeof target[prop] === "function") {
+      return function(...args) {
+        console.log(`'${prop} 'called with arguments: `, ...arguments);
+        return target[prop](...arguments);
+      };
+    }
+  }
+};
+
+const timerHandler = {
+  get: function(target, prop, receiver) {
+    if (typeof target[prop] === "function") {
+      return function(...args) {
+        console.log(`starting '${prop}'`);
+        const t1 = window.performance.now();
+
+        const res = target[prop](...arguments);
+        const t2 = window.performance.now();
+
+        console.log(`'${prop}' took ${t2 - t1}ns`);
+        return res;
+      };
+    }
+  }
+};
+
+const proxy = new Proxy(cow, tracerHandler);
+const proxy2 = new Proxy(proxy, timerHandler);
+console.log(proxy2.feed(20));
+console.log(proxy2.speak(3));
+```
+
+Unfortunately, in Go, there's no way to do this in a performant manner (the regular approach would involve using reflection and that can have a serious performance impact on the underlying code)
+
+## Our solution
+
+Implementation of decorator pattern that we chose involved 3 parts:
+
+1. Struct embedding
+2. Code parsing using AST
+3. Code generation using templates
+
+### Struct embedding
+
+Since Go is not object-oriented, it uses Struct embedding as a type of inheritance. Here's a small example:
+
+```go
+type Animal struct{
+	Name string
+}
+
+type Cow struct{
+	Animal
+}
+
+func (c Cow) Speak() {
+	fmt.Printf("Moo, I am a %s", c.Animal.Name)
+}
+
+func main() {
+	a := Animal{Name:"Cow"}
+	c := Cow{Animal:a}
+	c.Speak()
+}
+```
+
+How does struct embedding help us in implementation of decorator pattern?
+
+```go
+
+type IAnimal interface {
```

Comment:
Idiomatic Go suggests the name should be "Speaker". But 0/5 on what you think is appropriate.